rust/connlib/tunnel/src/peer.rs

```diff
@@ -508,6 +502,32 @@ impl ClientOnGateway {
     pub fn id(&self) -> ClientId {
         self.id
     }
+
+    /// Check all [`TranslationState`]s for dead but used IPs.
+    ///
+    /// We don't want to be spamming this warning but it also shouldn't go unnoticed.
+    /// Thus, the strategy is to only print the log if:
+    /// - An IP has recently been used (we have seen outgoing traffic in the last 30s)
+    /// - An IP has not responded in the last 10s
+    fn check_for_dead_ips(&self, now: Instant) {
+        let mut dead_ips = BTreeMap::<DomainName, BTreeSet<IpAddr>>::new();
+
+        for state in self.permanent_translations.values() {
+            if state.is_used(now) && state.is_dead(now) {
+                dead_ips
+                    .entry(state.name.clone())
+                    .or_default()
+                    .insert(state.resolved_ip);
+            }
+        }
+
+        if !dead_ips.is_empty() {
+            tracing::warn!(
```

Comment:
I wonder if this could be a bit spammy given multiple clients that are using this resource. What do you think about surfacing this another layer and coalesing these logs?

Reply:
Using an event you mean? I think we could! Do you think it is worth investing time into this now or first see how it actually pans out?

Reply:
I think the better fix here would be to introduce a shared DNS cache and do the detection only once instead of for each connection.